pilot/pkg/networking/core/v1alpha3/envoyfilter/listener_patch.go

```diff
@@ -126,9 +126,123 @@ func patchListener(patchContext networking.EnvoyFilter_PatchContext,
 			merge.Merge(listener, lp.Value)
 		}
 	}
+	patchListenerFilters(patchContext, patches[networking.EnvoyFilter_LISTENER_FILTER], listener)
 	patchFilterChains(patchContext, patches, listener)
 }
 
+// patchListenerFilters patches passed in filter if it is MERGE operation.
+// The return value indicates whether the filter has been removed for REMOVE operations.
+func patchListenerFilters(patchContext networking.EnvoyFilter_PatchContext,
+	patches []*model.EnvoyFilterConfigPatchWrapper,
+	listener *xdslistener.Listener,
+) {
+	removedFilters := sets.New[string]()
+	for _, lp := range patches {
+		if !commonConditionMatch(patchContext, lp) ||
+			!listenerMatch(listener, lp) {
+			IncrementEnvoyFilterMetric(lp.Key(), ListenerFilter, false)
+			continue
+		}
+		applied := false
+		if lp.Operation == networking.EnvoyFilter_Patch_ADD {
+			listener.ListenerFilters = append(listener.ListenerFilters, proto.Clone(lp.Value).(*xdslistener.ListenerFilter))
+			applied = true
+		} else if lp.Operation == networking.EnvoyFilter_Patch_INSERT_FIRST {
+			listener.ListenerFilters = append([]*xdslistener.ListenerFilter{proto.Clone(lp.Value).(*xdslistener.ListenerFilter)}, listener.ListenerFilters...)
+			applied = true
+		} else if lp.Operation == networking.EnvoyFilter_Patch_INSERT_AFTER {
+			// Insert after without a filter match is same as ADD in the end
+			if !hasListenerFilterMatch(lp) {
+				listener.ListenerFilters = append(listener.ListenerFilters, proto.Clone(lp.Value).(*xdslistener.ListenerFilter))
+				applied = true
+				continue
+			}
+
+			// find the matching filter first
```

Comment:
Can we refactor this logic in to func? We use it at multiple places to insert a filter at a position?

Reply:
I see that, good idea.  Need to dive into it

Reply:
This exists from beginnging, will clean it in other pr